contracts/modules/TransactionManager.sol

```diff
@@ -0,0 +1,226 @@
+    */
+    function isWhitelisted(
+        address _wallet,
+        address _target
+    )
+        public
+        view
+        returns (bool _isWhitelisted)
+    {
+        uint whitelistAfter = userWhitelist.getWhitelist(_wallet, _target);
+        return whitelistAfter > 0 && whitelistAfter < block.timestamp;
+    }
+
+    /** ******************* Callbacks ************************** */
+
+    /**
+    * @notice Implementation of EIP 1271.
+    * Should return whether the signature provided is valid for the provided data.
+    * @param _msgHash Hash of a message signed on the behalf of address(this)
+    * @param _signature Signature byte array associated with _msgHash
+    */
+    function isValidSignature(
+        bytes32 _msgHash,
+        bytes memory _signature
+    ) 
+        external
+        view
+        returns (bytes4)
+    {
+        require(_signature.length == 65, "TM: invalid signature length");
+        address signer = Utils.recoverSigner(_msgHash, _signature, 0);
+        require(_isOwner(msg.sender, signer), "TM: Invalid signer");
+        return ERC1271_ISVALIDSIGNATURE;
+    }
+
+    /**
+     * @notice Implementation of ERC721
+     * @notice An ERC721 smart contract calls this function on the recipient contract
+     * after a `safeTransfer`. If the recipient is a BaseWallet, the call to onERC721Received
+     * will be forwarded to the method onERC721Received of the present module.
+     * @return bytes4 `bytes4(keccak256("onERC721Received(address,address,uint256,bytes)"))`
+     */
+    function onERC721Received(
+        address /* operator */,
+        address /* from */,
+        uint256 /* tokenId */,
+        bytes calldata /* data*/
+    )
+        external
+        returns (bytes4)
+    {
+        return ERC721_RECEIVED;
+    }
+
+    // *************** Internal Functions ********************* //
+
+    function enableStaticCall(address _wallet) internal {
+        // setup static calls
+        IWallet(_wallet).enableStaticCall(address(this), ERC1271_ISVALIDSIGNATURE);
+        IWallet(_wallet).enableStaticCall(address(this), ERC721_RECEIVED);
+    }
+
+    function recoverSpender(address _wallet, Call calldata _transaction) internal pure returns (address) {
+        if (_transaction.isSpenderInData) {
+            require(_transaction.value == 0, "TM: unsecure call with spender in data");
+            // transfer(to, value), transferFrom(wallet, to, value),
+            (address first, address second) = abi.decode(_transaction.data[4:], (address, address));
+            return first == _wallet ? second : first;
+        }   
+        return _transaction.to;
+    }
+
+    function isAuthorised(address _wallet, address _spender, address _to, bytes calldata _data) internal view returns (bool) {
+        if (_to == _spender) {
+            return authoriser.authorise(_wallet, _to, _data); // do we need to block calls to the wallet or modules?
+        } else {
```

Comment:
Not sure this distinction is a good idea. If `spender != to`, then the only cases where a generic registry would be useful seems to be for ERC20 approvals. But in such cases, the filter would need to be able to distinguish an `approve()` call from another type of call. For example, a filter should deny a `transfer()` to the Paraswap Proxy but allow an `approve()` to it. Or a filter may allow any approval to the `UniswapV2Router` but only allow calls that do not use a beneficiary other than the wallet. So in general, the full `data` will be needed by the filter to make such distinctions. In conclusion I would only do `authoriser.authorise(_wallet, _to, _data);` here and let the filter figure out the details.

Reply:
Agreed.